validator/sawtooth_validator/journal/publisher.py

```diff
@@ -227,24 +245,37 @@ def finalize_block(self, identity_signing_key, pending_batches):
                     LOGGER.debug("Abandoning block %s:" +
                                  "root state hash has invalid txn applied",
                                  builder)
+                    # update the pending batch list to be all the
+                    # batches that passed validation to this point and
+                    # none of the ones that failed. It is possible that
+                    # this batch caused a future batch to fail so
+                    # we leave all of the batches that failed passed this
+                    # one in the list.
+                    bad_batches.append(batch)
+                    pending_batches = [x for x in self._pending_batches
+                                       if x not in bad_batches]
                     return None
                 else:
                     builder.add_batch(batch)
                     committed_txn_cache.add_batch(batch)
                 if result.state_hash is not None:
                     state_hash = result.state_hash
             else:
+                bad_batches.append(batch)
                 LOGGER.debug("Batch %s invalid, not added to block.",
                              batch.header_signature)
 
-        if state_hash is None:
+        if state_hash is None or not builder.batches:
             LOGGER.debug("Abandoning block %s no batches added", builder)
```

Comment:
Abandoning block %s -> Abandoning block %s:

Add a colon or some other punctuation to clarify that _"no batches added"_ is a separate clause, the reason for the abandonment.